File: go/src/deposit/src/create.go
package deposit

import (
	"errors"
	"os"
	"strings"
	"time"

	"interview_mock_deposits_go/deposit/src/domain"
	"interview_mock_deposits_go/utils/awsutil"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
)

var TableName = aws.String(os.Getenv("TABLE_NAME"))

var DdbSvc = awsutil.NewDynamoClient()

var ErrDepositAlreadyExists = errors.New("deposit already exists")

func CreateDeposit(request Transaction, requestId string) error {
	return createDeposit(request, nil, requestId, domain.New)
}

func CreateDepositWithUserData(request Transaction, userData UserData, requestId string) error {
	return createDeposit(request, &userData, requestId, domain.New)
}

func CreateDepositWithStatus(request Transaction, requestId string, status domain.DepositStatus) error {
	return createDeposit(request, nil, requestId, status)
}

func createDeposit(request Transaction, userData *UserData, requestId string, status domain.DepositStatus) error {
	newDeposit := Deposit{
		DepositId:   request.GenerateDepositId(),
		Status:      status,
		Transaction: request,
	}
	if userData != nil && userData.Document != "" {
		newDeposit.UserData = userData
	}
	return SaveDeposit(newDeposit, requestId)
}

func SaveDeposit(data Deposit, requestId string) error {
	item, err := dynamodbattribute.MarshalMap(data)
	if err != nil {
		return err
	}
	item["depositId"] = &dynamodb.AttributeValue{S: aws.String(data.DepositId)}
	item["createdAt"] = &dynamodb.AttributeValue{S: aws.String(time.Now().UTC().Format(time.RFC3339))}

	_, err = DdbSvc.PutItem(&dynamodb.PutItemInput{
		Item:                item,
		TableName:           TableName,
		ConditionExpression: aws.String("attribute_not_exists(depositId)"),
	})
	if err != nil && strings.Contains(err.Error(), "ConditionalCheckFailedException") {
		return ErrDepositAlreadyExists
	}
	return err
}

File: python/src/deposit/create.py
import os
from datetime import datetime, timezone
from typing import Any

from botocore.exceptions import ClientError

from deposit.domain.deposit_status import DepositStatus
from deposit.serde import deposit_to_ddb_item
from deposit.structs import Deposit, Transaction, UserData
from utils.awsutil import new_dynamo_client


class ErrDepositAlreadyExists(Exception):
    """Raised when a PutItem fails the depositId uniqueness condition."""


def _table_name() -> str:
    name = os.environ.get("TABLE_NAME")
    if not name:
        raise RuntimeError("TABLE_NAME env var is required")
    return name


def _ddb() -> Any:
    return new_dynamo_client()


def create_deposit(request: Transaction, request_id: str) -> None:
    _create(request, None, request_id, DepositStatus.NEW)


def create_deposit_with_user_data(request: Transaction, user_data: UserData, request_id: str) -> None:
    _create(request, user_data, request_id, DepositStatus.NEW)


def create_deposit_with_status(request: Transaction, request_id: str, status: DepositStatus) -> None:
    _create(request, None, request_id, status)


def _create(
    request: Transaction,
    user_data: UserData | None,
    request_id: str,
    status: DepositStatus,
) -> None:
    new = Deposit(
        deposit_id=request.generate_deposit_id(),
        status=status,
        transaction=request,
    )
    if user_data is not None and user_data.document:
        new.user_data = user_data
    save_deposit(new, request_id)


def save_deposit(data: Deposit, request_id: str) -> None:
    item = deposit_to_ddb_item(data)
    item["createdAt"] = {"S": datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")}
    try:
        _ddb().put_item(
            TableName=_table_name(),
            Item=item,
            ConditionExpression="attribute_not_exists(depositId)",
        )
    except ClientError as exc:
        if exc.response.get("Error", {}).get("Code") == "ConditionalCheckFailedException":
            raise ErrDepositAlreadyExists("deposit already exists") from exc
        raise
